design-proposals/klip-27-enhanced-udf-configuration-options.md

```diff
@@ -0,0 +1,111 @@
+
+## Design
+
+A simple approach is to extend the function meta-data and add a _groups_ option to the `@UdfDecription` annotation, which takes a list of strings. Each string entry in _groups_ explicitly defines a name used to logically group functions for which to share configuration properties.
+
+Given the following example:
+
+```java
+@UdfDescription(
+    name = "someudf",
+    description = "...",
+    author = "...",
+    version = "...",
+    groups = {"shareit"/*,"..."*/})
+public class SomeUdf implements Configurable {
+    //...
+}
+```
+
+allows to configure the function _someudf_ using `ksql.functions.someudf.my.value.x=foo` (basic, single-UDF config) or `ksql.groups.functions.shareit.my.value.x=foo` (grouped config across several functions declaring the same group name(s)). This distinction also makes it obvious, whether configuration properties are set for single functions or several grouped functions respectively.
+
+A drawback of the meta-data driven approach is that the group information must be declared in the code. If this restriction turns out to be an issue, a future enhancement could allow to explicitly define the group information in the configuration properties like so:  ?`ksql.groups.define.group_name=...,...` which for the particular example above would read `ksql.groups.define.shareit=someudf`.
+
+The benefits of this design are:
+
+* full backwards compatibility because groups config meta-data is a new option
+* the default and most common case is to configure a single function which means that it stays simple for users
+* it is easy to add new configurations
+* the grouping is neither derived from the name itself (e.g. prefix matching) nor is it tied to java package naming convention
+
+## Test plan
+
+Test and failure scenarios are rather straight-foward, focusing only on different ways to set and read configuration properties for UDFs. There shouldn't be a need for scale/load/performance testing because this KLIP doesn't affect any criticial areas in that regard.
+
+## LOEs and Delivery Milestones
+
+To be defined by product manager and team.
+
+## Documentation Updates
+
+The current documentation regarding configuration options for custom UDFs isn't overly specific but of course needs to be updated accordingly. **It's important to be very clear and call out the dropped support of the `_global_` configuration scope which leads to a breaking change for upgrading/migrating workloads.**
+
+## Compatibility Implications
+
+As the design proposal suggests to remove support of the `_global_` scope, this means that upgrading the ksqlDB version will lead to a breaking change in that regard. However, the implementation of the `*` wildcard prefix matching of UDFs' names should be carried out in a backwards compatible way such that it can reasonably support the "old" (i.e. current) and the "new" (i.e. what's proposed here) configuration options in parallel.
```

Comment:
this seems no longer accurate either (`*` is not being introduced). With this newer approach, can we just make it backwards compatible and keep the `_global_` scope?

Reply:
thx. turns out I'm a bit blind when reading this for the ... time ;-)